File: py/median_of_two_sorted_arrays.py
# -*- coding: utf-8 -*-
from typing import *

class Solution:
    def findMedianSortedArrays(self, a: List[int], b: List[int]) -> float:
        class SafeList(list):
            def __getitem__(self, idx):
                if idx < 0:
                    return float("-inf")
                try:
                    return super().__getitem__(idx)
                except:
                    return float("inf")

        a, b = sorted((a, b), key=len)
        a, b = SafeList(a), SafeList(b)
        m, n = len(a), len(b)
        after = (m + n + 1) // 2

        lo, hi = 0, m
        while lo < hi:
            pa = (lo + hi) // 2
            if a[pa] >= b[after - pa - 1]:
                hi = pa
            else:
                lo = pa + 1
        pa = lo
        pb = after - pa

        max_left_a, min_right = a[pa - 1], a[pa]
        max_left_b, min_right_b = b[pb - 1], b[pb]
        return (
            (max(max_left_a, max_left_b) + min(min_right, min_right_b)) / 2
            if not (m + n) % 2
            else max(max_left_a, max_left_b)
        )

File: java/medianOfTwoSortedArrays.java
class Solution {
    private int safeGet(int[] v, int idx) {
        if (idx < 0)
            return Integer.MIN_VALUE;
        if (idx >= v.length)
            return Integer.MAX_VALUE;
        return v[idx];
    }

    public double findMedianSortedArrays(int[] a, int[] b) {
        int m = a.length, n = b.length;
        int lo = 0, hi = m, after = (m + n + 1) / 2, pa;
        while (lo < hi) {
            pa = lo + (hi - lo) / 2;
            if (a[pa] >= this.safeGet(b, after - pa - 1))
                hi = pa;
            else
                lo = pa + 1;
        }
        pa = lo;
        int pb = after - pa;
        double maxLeft = Math.max(this.safeGet(a, pa - 1), this.safeGet(b, pb - 1));
        double minRight = Math.min(this.safeGet(a, pa), this.safeGet(b, pb));
        return (m + n) % 2 == 1 ? maxLeft : (maxLeft + minRight) / 2;
    }
    
    public static void main(String[] args) {
        int a[] = { 1, 2, 3, 4, 5, };
        int b[] = { 3, 4, 5, 6, };
        Solution s = new Solution();
        System.out.println(s.findMedianSortedArrays(a, b));
    }
}